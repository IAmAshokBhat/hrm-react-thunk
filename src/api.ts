import {
  ILoginData,
  INewAttendance,
  INewLeave,
  IRegisterAPI
} from './constants';
import { fetchWithToken, getToken } from './utils';

const headers = {
  'X-JWT-Token': getToken()
};

export const getUserListAPI = () => fetchWithToken('user/list');

export const getUserDetailsAPI = (id: number) =>
  fetchWithToken(`api/users/${id}`);

export const getAttendanceDetailsAPI = (id: number, from: string, to: string) =>
  fetchWithToken(`api/attendance/${id}?from=${from}&to=${to}`);

export const getLeaveBalanceAPI = (id: number) =>
  fetchWithToken(`api/leaves/${id}`);

export const getEventDetailsAPI = (id: number, from: string, to: string) =>
  fetchWithToken(`api/events/${id}?from=${from}&to=${to}`);

export const getAllLeavesAPI = (id: number) =>
  fetchWithToken(`api/all-leaves/${id}`);

export const applyLeaveAPI = (newLeave: INewLeave) =>
  fetchWithToken(`api/apply-leave`, {
    method: 'POST',
    body: JSON.stringify(newLeave)
  });

export const addAttendanceAPI = (newAttendance: INewAttendance) =>
  fetchWithToken(`api/attendance`, {
    method: 'POST',
    body: JSON.stringify(newAttendance)
  });

export const loginAPI = (userInfo: ILoginData) =>
  fetchWithToken(`login`, {
    method: 'POST',
    body: JSON.stringify(userInfo)
  });

export const getDesignationListAPI = () =>
  fetchWithToken(`designation`, { headers: headers });

export const registerAPI = (newUser: IRegisterAPI) =>
  fetchWithToken(`user`, {
    method: 'POST',
    body: JSON.stringify(newUser)
  });
