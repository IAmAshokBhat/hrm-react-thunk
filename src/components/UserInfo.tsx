import Avatar from '@mui/material/Avatar';
import Typography from '@mui/material/Typography';
import Card from '@mui/material/Card';
import Grid from '@mui/material/Grid';
import { useDispatch, useSelector } from 'react-redux';
import { fetchUserDetailsAction } from '../redux/actions';
import { useEffect } from 'react';
import { selectLoginDetails, selectUserInfo } from '../redux/selectors';

export const UserInfo = () => {
  const dispatch = useDispatch();

  const userInfo = useSelector(selectUserInfo);

  const loginDetails = useSelector(selectLoginDetails);

  let userDetails = {
    name: '',
    userId: -1,
    designation: '',
    manager: '',
    email: ''
  };

  if (userInfo) {
    const { name, userId, designation, manager, email } = userInfo;
    userDetails = {
      name,
      userId,
      designation,
      manager,
      email
    };
  }

  useEffect(() => {
    dispatch(fetchUserDetailsAction(loginDetails.userId));
  }, [dispatch, loginDetails]);

  return (
    <Card style={{ padding: 15 }}>
      <Grid container>
        <Avatar
          alt={userDetails.name}
          sx={{ width: 156, height: 156, bgcolor: '#2976d2' }}
          style={{ left: 100 }}
        />
        <Grid item xs={12}>
          <Typography variant="h5">{userDetails.name}</Typography>
          <div>
            <Typography variant="subtitle1" component="span">
              {userDetails.designation}
            </Typography>
          </div>
          <div>
            <Typography variant="subtitle1" component="span" fontWeight={500}>
              Email:
            </Typography>
            <Typography variant="subtitle1" component="span">
              {userDetails.email}
            </Typography>
          </div>
          <div>
            <Typography variant="subtitle1" component="span" fontWeight={500}>
              Manager:
            </Typography>
            <Typography variant="subtitle1" component="span">
              {userDetails.manager}
            </Typography>
          </div>
          <div>
            <Typography variant="subtitle1" component="span" fontWeight={500}>
              Employee ID:
            </Typography>
            <Typography variant="subtitle1" component="span">
              {userDetails.userId}
            </Typography>
          </div>
        </Grid>
      </Grid>
    </Card>
  );
};
