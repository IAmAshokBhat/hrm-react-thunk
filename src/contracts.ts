import { AnyAction, Dispatch } from 'redux';
import { TAny, TAnyObject } from './constants';

export type TFPromise = (params: TAny) => Promise<TAnyObject>;

export type TFPromiseAPI = (params?: TAny | TAnyObject) => Promise<TAnyObject>;

export type TFAction = (dispatch: Dispatch<AnyAction>) => Promise<void>;

/*
()
(1,2,3)
({key:value})
**
*/
