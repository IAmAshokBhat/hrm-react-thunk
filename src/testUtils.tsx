import { render, RenderOptions, waitFor } from '@testing-library/react';
import { SnackbarProvider } from 'notistack';
import { ReactElement } from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { applyMiddleware, createStore, Store } from 'redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { IDefaultRoot, TAny, TAnyObject } from './constants';
import { TFAction } from './contracts';
import { reducers } from './redux/reducers';
import {
  defaultArrayState,
  defaultNullState,
  defaultState
} from './redux/reducers/__tests__/fixtures';

export const middlewares = [thunk];

export const createStoreWithMiddlewares = (
  initialState: TAnyObject = {}
): Store => {
  return createStore(reducers, initialState, applyMiddleware(...middlewares));
};
interface IRenderWithProviderOptions {
  initialStoreState?: TAnyObject;
  renderFunction?: TAny;
  options?: RenderOptions;
  route?: string;
  path?: string;
}

export const renderWithProvider = (
  ui: ReactElement,
  renderOptions?: IRenderWithProviderOptions
) => {
  const {
    initialStoreState = {},
    renderFunction = render,
    options = {},
    route = '',
    path = ''
  } = renderOptions || {};

  const appStore = createStoreWithMiddlewares(initialStoreState);

  let wrappedUi = (
    <SnackbarProvider
      anchorOrigin={{
        vertical: 'top',
        horizontal: 'right'
      }}
    >
      <Provider store={appStore}>{ui}</Provider>
    </SnackbarProvider>
  );

  if (route) {
    window.history.pushState({}, '', route);

    wrappedUi = (
      <Provider store={appStore}>
        <BrowserRouter>
          <Routes>
            <Route path={path} element={ui} />
          </Routes>
        </BrowserRouter>
      </Provider>
    );
  }

  return renderFunction(wrappedUi, options);
};

type TFApiCall = (...params: TAny) => Promise<Response>;

export interface IResponseMock {
  headers?: object;
  ok?: boolean;
  status?: number;
  statusText?: string;
  type?: ResponseType;
  url?: string;
  body?: TAny | null;
  bodyUsed?: boolean;
  blob?: () => Promise<TAny>;
}

const dummyPromise = () => new Promise((resolve) => resolve('dummy'));

export class ResponseMock {
  headers: object;

  ok: boolean;

  status: number;

  statusText: string;

  type: ResponseType;

  url: string;

  body: ReadableStreamDefaultReadResult<Uint8Array> | null;

  bodyUsed: boolean;

  blob: () => Promise<TAny>;

  constructor(props: IResponseMock = {}) {
    this.headers = props.headers || {};
    this.status = props.status || 200;
    this.ok = props.ok || this.status === 200;
    this.statusText = props.statusText || 'OK';
    this.type = props.type || 'basic';
    this.url = props.url || 'http://localhost';
    this.body = props.body;
    this.bodyUsed = props.bodyUsed || this.body !== null;
    this.blob = props.blob || dummyPromise;
  }

  json: () => Promise<TAny> = () => {
    const body = this.body || '""';

    try {
      return Promise.resolve(JSON.parse(body as string));
    } catch {
      return Promise.resolve(body);
    }
  };

  text: () => Promise<string> = () => Promise.resolve(this.statusText);
}

export const mockAPICall = (func: TFApiCall, response: IResponseMock) =>
  (func as jest.Mock).mockResolvedValue(new ResponseMock(response));

export const generateActionTests = async (
  actionName: string,
  action: TFAction,
  api: TAny,
  actionPayload: TAny = []
) => {
  const response = new ResponseMock({ ok: true, body: { data: [] } });

  const mockStore = configureStore(middlewares);

  (api as jest.Mock).mockReturnValue(Promise.resolve(response));
  const store = mockStore({});

  //TODO: Fix TS error
  // @ts-ignore
  await store.dispatch(action);

  await waitFor(() => {
    expect(store.getActions()).toEqual([
      {
        type: `${actionName}_LOADING`
      },
      {
        type: `${actionName}_SUCCESS`,
        //TODO: Accept action params from function params
        payload: actionPayload
      }
    ]);
  });
};

export const generateReducerTests = (
  reducerName: string,
  reducer: TAny,
  selector: TAny,
  actionName: string,
  successValue?: TAny,
  defaultValue?: TAny,
  expectedValue?: TAny,
  reducerLabel?: string
) => {
  describe(`for ${reducerName}`, () => {
    const reducerLbl = reducerLabel
      ? reducerLabel
      : reducerName.replace('Reducer', '');
    it(`${reducerName} reducer loading`, () => {
      reducer(undefined, {
        type: `${actionName}_LOADING`,
        payload: null
      });
      // TODO: Check why its failing
      // expect(selector(defaultState).loading).toEqual(true);
      expect(selector(defaultState).value).toEqual(
        defaultValue === null ? defaultValue : []
      );
      expect(selector(defaultState).error).toEqual(null);
    });

    it(`${reducerName} reducer success`, () => {
      const successState: IDefaultRoot = {
        ...defaultState,
        [reducerLbl]: {
          ...defaultNullState,
          value: expectedValue ? expectedValue : successValue
        }
      };
      reducer(defaultArrayState, {
        type: `${actionName}_SUCCESS`,
        payload: successValue
      });

      expect(selector(successState).error).toEqual(null);
      expect(selector(successState).loading).toEqual(false);
      expect(selector(successState).value).toEqual(
        expectedValue ? expectedValue : successValue
      );
    });

    it(`${reducerName} reducer error`, () => {
      const error = { message: '', status: 1, ok: false };
      const failedState: IDefaultRoot = {
        ...defaultState,
        [reducerLbl]: {
          ...defaultNullState,
          value: [],
          error
        }
      };

      reducer(undefined, {
        type: `${actionName}_FAILURE`,
        payload: error
      });

      expect(selector(failedState).loading).toEqual(false);
      expect(selector(failedState).value).toEqual([]);
      expect(selector(failedState).error).toEqual(error);
    });
  });
};
