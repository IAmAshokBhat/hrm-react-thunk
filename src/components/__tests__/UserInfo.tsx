import { getUserDetailsAPI } from '../../api';
import { TAnyObject } from '../../contracts';
import { mockAPICall, renderWithProvider } from '../../testUtils';
import { UserInfo } from '../UserInfo';
import { defaultState, userInfoAPIResponse } from '../__tests__/fixtures';

jest.mock('../../api', () => ({
  getUserDetailsAPI: jest.fn()
}));

const userInfoResponse = {
  ok: true,
  body: userInfoAPIResponse
};

describe('User Info component', () => {
  const generateNode = (initialStoreState: TAnyObject = defaultState) =>
    renderWithProvider(<UserInfo />, {
      route: `/dashboard`,
      path: '/dashboard',
      initialStoreState
    });

  beforeEach(() => {
    mockAPICall(getUserDetailsAPI, userInfoResponse);
  });

  it('should render and match snapshot', () => {
    const { container } = generateNode();

    expect(container).toMatchSnapshot();
  });
});
