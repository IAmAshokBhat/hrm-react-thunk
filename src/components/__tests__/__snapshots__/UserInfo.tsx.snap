// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Info component should render and match snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
    style="padding: 15px;"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1cljjma-MuiAvatar-root"
        style="left: 100px;"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
          data-testid="PersonIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
          />
        </svg>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        >
          ABC
        </h5>
        <div>
          <span
            class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
          >
            Software Engineer
          </span>
        </div>
        <div>
          <span
            class="MuiTypography-root MuiTypography-subtitle1 css-e0r9sn-MuiTypography-root"
          >
            Email:
          </span>
          <span
            class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
          >
            noemail@email.com
          </span>
        </div>
        <div>
          <span
            class="MuiTypography-root MuiTypography-subtitle1 css-e0r9sn-MuiTypography-root"
          >
            Manager:
          </span>
          <span
            class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
          >
            XYZ
          </span>
        </div>
        <div>
          <span
            class="MuiTypography-root MuiTypography-subtitle1 css-e0r9sn-MuiTypography-root"
          >
            Employee ID:
          </span>
          <span
            class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
          >
            1
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
